contracts/HATVault.sol

```diff
@@ -0,0 +1,1109 @@
+        } else {
+            return registry.defaultChallengeTimeOutPeriod();
+        }
+    }
+
+    /* -------------------------------------------------------------------------------- */
+
+    /* --------------------------------- Helpers -------------------------------------- */
+
+    /**
+    * @dev Deposit funds to the vault. Can only be called if the committee had
+    * checked in and deposits are not paused.
+    * NOTE: Vaults should not use tokens which do not guarantee that the 
+    * amount specified is the amount transferred
+    * @param caller Caller of the action (msg.sender)
+    * @param receiver Reciever of the shares from the deposit
+    * @param assets Amount of vault's native token to deposit
+    * @param shares Respective amount of shares to be received
+    */
+    function _deposit(
+        address caller,
+        address receiver,
+        uint256 assets,
+        uint256 shares
+    ) internal override virtual nonReentrant {
+        if (!committeeCheckedIn)
+            revert CommitteeNotCheckedInYet();
+        if (shares == 0) revert AmountToDepositIsZero();
+        if (withdrawEnableStartTime[receiver] != 0 && receiver == caller) {
+            // clear withdraw request
+            withdrawEnableStartTime[receiver] = 0;
+        }
+
+        super._deposit(caller, receiver, assets, shares);
+    }
+
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares,
+        uint256 fee
+    ) internal nonReentrant {
+        if (assets == 0) revert WithdrawMustBeGreaterThanZero();
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        _burn(owner, shares);
+
+        IERC20 asset = IERC20(asset());
+        if (fee > 0) {
+            asset.safeTransfer(registry.owner(), fee);
+        }
+        asset.safeTransfer(receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal virtual override {
+        if (to != address(0)) {
+            if (registry.isEmergencyPaused()) revert SystemInEmergencyPause();
+            // Cannot transfer or mint tokens to a user for which an active withdraw request exists
+            // because then we would need to reset their withdraw request
+            if (withdrawEnableStartTime[to] != 0) {
+                HATVaultsRegistry.GeneralParameters memory generalParameters = registry.getGeneralParameters();
+                // solhint-disable-next-line not-rely-on-time
+                if (block.timestamp < withdrawEnableStartTime[to] + generalParameters.withdrawRequestEnablePeriod)
+                    revert CannotTransferToAnotherUserWithActiveWithdrawRequest();
+            }
+            rewardController.updateVaultBalance(to, amount, true);
```

Comment:
If this returns an error, then all transfers (and deposits and withdrawals) will revert

this is too much power. Can we do a try . catch here?

Reply:
Yes, but then we risk the rewardController getting out of sync. Not sure what the best solution would be here

Reply:
Currently it means that whoever can call the setRewardController can freeze the funds which is bad, maybe we should do with try only for withdraw? Or maybe add a delay on changing the reward controller?

Reply:
Or even just remove setRewardController